File: app/services/conversation/chat_session_repository.rb
class Conversation::ChatSessionRepository
  def initialize(config: AppConfig.fetch)
    @config = config
  end

  def find_active(chat_id)
    session = ChatSession.find_by(chat_id: chat_id)
    return nil if session.nil?

    return session if current_time_ms - session.updated_at <= session_ttl_ms

    ChatSession.where(chat_id: chat_id).delete_all
    Rails.logger.info("Reset expired session chat_id=#{chat_id}")
    nil
  end

  def persist(chat_id, conversation_state)
    record = ChatSession.find_or_initialize_by(chat_id: chat_id)
    record.last_response_id = conversation_state
    record.updated_at = current_time_ms
    record.save!
  end

  def reset(chat_id)
    ChatSession.where(chat_id: chat_id).delete_all
    Rails.logger.info("Reset chat session chat_id=#{chat_id}")
  end

  private

  def session_ttl_ms
    @config.session_ttl_days * 24 * 60 * 60 * 1000
  end

  def current_time_ms
    (Time.now.to_f * 1000).to_i
  end
end

File: src/main/java/com/telegram/codex/conversation/session/ChatSessionRepository.java
package com.telegram.codex.conversation.session;

import com.telegram.codex.config.AppProperties;
import com.telegram.codex.persistence.ChatSessionEntity;
import com.telegram.codex.persistence.ChatSessionJpaRepository;
import java.time.Duration;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Repository;

@Repository
public class ChatSessionRepository {

    private static final Logger LOGGER = LoggerFactory.getLogger(ChatSessionRepository.class);

    private final AppProperties properties;
    private final ChatSessionJpaRepository repository;

    public ChatSessionRepository(AppProperties properties, ChatSessionJpaRepository repository) {
        this.properties = properties;
        this.repository = repository;
    }

    public Optional<ChatSessionRecord> findActive(String chatId) {
        Optional<ChatSessionEntity> maybeSession = repository.findById(chatId);
        if (maybeSession.isEmpty()) {
            return Optional.empty();
        }
        ChatSessionEntity entity = maybeSession.get();
        if (currentTimeMs() - entity.getUpdatedAt() > sessionTtlMs()) {
            repository.deleteById(chatId);
            LOGGER.info("Reset expired session chat_id={}", chatId);
            return Optional.empty();
        }
        return Optional.of(toRecord(entity));
    }

    public void persist(String chatId, String conversationState) {
        ChatSessionEntity entity = repository.findById(chatId).orElseGet(ChatSessionEntity::new);
        entity.setChatId(chatId);
        entity.setLastResponseId(conversationState);
        entity.setUpdatedAt(currentTimeMs());
        repository.save(entity);
    }

    public void reset(String chatId) {
        repository.deleteById(chatId);
        LOGGER.info("Reset chat session chat_id={}", chatId);
    }

    private long sessionTtlMs() {
        return Duration.ofDays(properties.getSessionTtlDays()).toMillis();
    }

    private long currentTimeMs() {
        return System.currentTimeMillis();
    }

    private ChatSessionRecord toRecord(ChatSessionEntity entity) {
        return new ChatSessionRecord(entity.getChatId(), entity.getLastResponseId(), entity.getUpdatedAt());
    }
}
